hpx/lcos/detail/future_data.hpp

```diff
@@ -489,10 +496,20 @@ namespace detail
         /// operation. Allows any subsequent set_data operation to succeed.
         void reset(error_code& /*ec*/ = throws)
         {
-            boost::unique_lock<mutex_type> l(this->mtx_);
+            state s = empty;
+
+            {
+                boost::unique_lock<mutex_type> l(this->mtx_);
+
+                std::swap(state_, s);
+                on_completed_ = completed_callback_type();
+
+                // unlock for the destruction of embedded data
+                // FIXME: is that correct?
+            }
```

Comment:
Ok, but the we don't need the lock in the first place. This lock was causing the issues when reset was invoked from the destructor.